\documentclass[10pt, british]{beamer}
\usetheme{UniBo}

\usefonttheme{professionalfonts} % prevents beamer from overriding fonts
\setmonofont{Fira Code Regular}

\addbibresource{./references.bib}

\title[Single Aminoacid Variants Prediction]{Prediction of the Effect of Single Amino Acid Protein
	Variants\\Using Deep Mutational Scanning Data}
\subtitle[]{University of Bologna --- Master Thesis in Bioinformatics}
\author[Pierotti Saul]{Pierotti Saul}
\advisor[]{%
	\setlength{\tabcolsep}{0pt}%
	\begin{tabular}{llll}
		Internal~ & Advisor:~ & Prof.~ & Pietro Di Lena                       \\
		External~ & Advisor:~ & Prof.~ & Arne Elofsson (Stockholm University)
	\end{tabular}
}
\date{July 19, 2021}

\begin{document}

\begin{frame}[plain]
	\titlepage%
\end{frame}

\begin{frame}
	\frametitle{Deep mutational scanning}
	High-throughput technique for obtaining fitness information on a large number of mutations
	\vfill%
	\begin{figure}
		\input{tikz/dms_workflow.tex}
	\end{figure}
\end{frame}

\begin{frame}
	\frametitle{I considered only single aminoacid variants}
	\vfill%
	\begin{figure}
		\input{tikz/mutation_illustration.tex}
	\end{figure}
\end{frame}

%\begin{frame}
%	\frametitle{Effect of single amino acid substitutions}
%	\begin{figure}
%		\input{tikz/effect_of_mutations_mindmap.tex}
%	\end{figure}
%\end{frame}

%\begin{frame}
%	\frametitle{Variant effect prediction}
%	Complete experimental coverage of the human proteome mutational landscape is not currently in reach, and it is a distant goal for model organisms.
%	Computational predictions are needed.
%
%	But why is it important to know the effect of a mutation?
%
%	\begin{itemize}
%		\item Precision medicine
%		\item Protein engineering
%	\end{itemize}
%
%	How can it be done?
%
%	\begin{itemize}
%		\item Supervised or unsupervised
%		\item Quantitative or as a classification problem
%		\item Machine learning or simple statistical models
%	\end{itemize}
%\end{frame}

\begin{frame}
	\frametitle{Some notable variant effect predictors}
	\begin{figure}
		\input{tikz/variant_effect_predictors.tex}
	\end{figure}
	\vfill%
	Among these predictors only Envision was trained on deep mutational scanning data
\end{frame}

\begin{frame}
	\frametitle{Structure of the project}
	\begin{figure}
		\input{tikz/project_structure.tex}
	\end{figure}
\end{frame}

\begin{frame}
	\frametitle{I used the training dataset of Envision \parencite{Gray2018}}
	\vfill%
	\centering%
	% for some reason figure fails
	\input{tikz/tikzdevice/presentation_fitness_score_distribution.tex}
	\input{tikz/tikzdevice/presentation_datasets_abundancies.tex}
\end{frame}

\begin{frame}
	\frametitle{Poor correlation among experimental results}
	Two independent deep mutational scanning experiments on Ubiquitin are present in the training dataset.
	Their correlation is low.
	\vfill%
	\centering%
	\input{tikz/tikzdevice/presentation_datasets_correlations.tex}
\end{frame}

\begin{frame}
	\frametitle{The identities of the mutated residues are important}
	Mutations from polar residues are less detrimental than other mutations.
	This effect disappears when filtering by relative solvent accessibility (not shown here).
	\vfill%
	\centering%
	\input{tikz/tikzdevice/presentation_mutation_identity.tex}
\end{frame}

\begin{frame}
	\frametitle{Buried residues are more conserved}
	\centering%
	\input{tikz/tikzdevice/presentation_score_dssp_rsa.tex}
\end{frame}

\begin{frame}
	\frametitle{Secondary structure is not critical}
	\centering%
	\input{tikz/tikzdevice/presentation_score_dssp_q3.tex}
\end{frame}

\begin{frame}
	\frametitle{I did not use structural information}
	\centering%
	\begin{figure}
		\input{tikz/features.tex}
	\end{figure}
\end{frame}

\begin{frame}
	\frametitle{Single protein models with naive testing}
	\begin{itemize}
		\item Half of the mutations in a protein randomly set aside for testing
		\item Hyperparameters optimized in the remaining half with 5-fold cross-validation
		\item Good results but likely overfitting the testing set
	\end{itemize}
	\vfill%
	\centering%
	\input{tikz/tikzdevice/presentation_single_protein_models_naive_test_result_summary.tex}
\end{frame}

\begin{frame}
	\frametitle{Single protein models with testing by position}
	\begin{itemize}
		\item Half of the mutations in a protein set aside for testing but avoiding mutations in the same protein position to end up in different splits
		\item Hyperparameters optimized in the remaining half with 5-fold cross-validation
		\item Performance more realistic
	\end{itemize}
	\vfill%
	\centering%
	\input{tikz/tikzdevice/presentation_single_protein_models_by_position_test_result_summary.tex}
\end{frame}

\begin{frame}
	\frametitle{Leave-One-Protein-Out (LOPO) models}
	\begin{itemize}
		\item For the left-out protein, half of the mutations used for testing
		\item Spearman correlation coefficient used for evaluation
		\item Small difference among gradient boosted trees and linear regression
		\item Performances comparable to those of Envision
	\end{itemize}
	\vfill%
	\centering%
	\input{tikz/tikzdevice/presentation_performance_comparison.tex}
\end{frame}

\begin{frame}
	\frametitle{Discussion}
	\begin{figure}
		\begin{tikzpicture}
			\tikzstyle{every node}=[object, font=\footnotesize]
			\node [object, text width=10em] at (0,0) {How validation and testing are performed is crucial};
			\node [object, text width=14em] at (5,-1) {Performances on par with other predictors can be reached without structural features};
			\node [object, text width=9em] at (1,2) {There is strong variability between datasets};
			\node [object, text width=10em] at (0,4) {Complex models do not improve much on linear regression};
			\node [object, text width=10em] at (6,3) {Unsupervised models perform similarly to supervised models};
		\end{tikzpicture}
	\end{figure}
\end{frame}

\begin{frame}
	\frametitle{Future directions}
	\begin{figure}
		\begin{tikzpicture}
			\tikzstyle{every node}=[object, font=\footnotesize]
			\node [object, text width=14em] at (0,0) {Using residue contacts in a graph convolutional neural network};
			\node [object] at (5,1) {Trying different models};
			\node [object, text width=7em] at (1,2) {Tuning the set of features};
			\node [object, text width=10em] at (6,3) {Training on more deep mutational scanning studies};
			\node [object, text width=13em] at (1,4) {Unsupervised models seem promising and may be worth exploring more};
			\node [object, text width=15em] at (6,-1) {Finding a better normalization strategy for the scores from different experiments};
		\end{tikzpicture}
	\end{figure}
\end{frame}

\appendix%
\begin{frame}[allowframebreaks]
	\frametitle{Bibliography}
	\printbibliography{}%
\end{frame}

\end{document}
