\currentpdfbookmark{Abstract}{abstract}%
\null\cleardoublepage%
\begin{abstract}
	Knowledge of the effect of mutations is crucial for improving our understanding of protein function.
	Moreover, the rapid increase in the availability of genomic data poses the challenge of linking uncharacterised genetic variants to phenotypes.
	Deep mutational scanning is an experimental mutagenesis technique that leverages the use of next-generation sequencing for the assessment of the effect of mutations.
	This approach is high-throughput compared to site-directed mutagenesis, but still, it is not sufficient for the characterization of the mutational landscape of the known proteomes.
	As a consequence, computational tools for the prediction of the effect of mutations are a valuable resource.
	In this work, I present a series of supervised regression models for the prediction of the quantitative effect of mutations trained on deep mutational scanning data.
	My models, despite not requiring structural information, perform similarly to Envision, a predictor that uses structural features.
	I compare gradient boosted trees and linear regression models, and I also explore several validation and testing strategies.
\end{abstract}
\null\cleardoublepage%
\currentpdfbookmark{\contentsname}{toc}%
{%
	\hypersetup{linkcolor=.}%
	\tableofcontents%
	\listoffigures%
	\addcontentsline{toc}{chapter}{List of Figures}%
	\listoftables%
	\addcontentsline{toc}{chapter}{List of Tables}%
}%
