{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "reserved-challenge",
   "metadata": {},
   "source": [
    "# Dataset preparation humsavar\n",
    "I use the humsavar dataset consisting in single aminoacid variants from uniprot. This is the same dataset used in Savojardo2021. I can also compare it with the datasets used in Kvist2018."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "durable-characterization",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--------------------------------------------------------------------------------\n",
      "        UniProt - Swiss-Prot Protein Knowledgebase\n",
      "        SIB Swiss Institute of Bioinformatics; Geneva, Switzerland\n",
      "        European Bioinformatics Institute (EBI); Hinxton, United Kingdom\n",
      "        Protein Information Resource (PIR); Washington DC, USA\n",
      "--------------------------------------------------------------------------------\n",
      "\n",
      "Description: Index of human variants curated from literature reports\n",
      "Name:        humsavar.txt\n",
      "Release:     2021_01 of 10-Feb-2021\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "This file lists all missense variants annotated in UniProtKB/Swiss-Prot human\n",
      "entries. It provides a variant classification which is intended for research\n",
      "purposes only, not for clinical and diagnostic use.\n",
      "\n",
      " - The column 'Variant category' shows the classification of the variant using\n",
      "   the American College of Medical Genetics and Genomics/Association for\n",
      "   Molecular Pathology (ACMG/AMP) terminology (Richards et al. PubMed:25741868)\n",
      "   into the following categories:\n",
      "\n",
      "   LP/P = likely pathogenic or pathogenic\n",
      "   LB/B = likely benign or benign\n",
      "   US   = uncertain significance\n",
      "\n",
      "   These categories are assigned based on the variant annotation in the\n",
      "   corresponding UniProtKB/Swiss-Prot entries that is curated from literature\n",
      "   reports. The classification may change over time and must not be considered\n",
      "   as a definitive statement about the pathogenic role of a variant.\n",
      "\n",
      " - The column 'Disease name' shows the name of the disease or the disease sample\n",
      "   in which variants have been found. Names are only provided for diseases\n",
      "   catalogued in OMIM and for cancer samples.\n",
      "\n",
      "Statistics for single amino acid variants:\n",
      "\n",
      "             LP/P:  31218\n",
      "             LB/B:  39527\n",
      "             US:    8631\n",
      "             Total: 79376\n",
      "\n",
      "Main      Swiss-Prot             AA             Variant\n",
      "gene name AC         FTId        change         category dbSNP          Disease name\n",
      "_________ __________ ___________ ______________ ________ ______________ _____________________\n",
      "A1BG      P04217     VAR_018369  p.His52Arg     LB/B     rs893184       -\n",
      "A1BG      P04217     VAR_018370  p.His395Arg    LB/B     rs2241788      -\n",
      "A1CF      Q9NQ94     VAR_052201  p.Val555Met    LB/B     rs9073         -\n",
      "A1CF      Q9NQ94     VAR_059821  p.Ala558Ser    LB/B     rs11817448     -\n",
      "A2M       P01023     VAR_000012  p.Arg704His    LB/B     rs1800434      -\n",
      "A2M       P01023     VAR_000013  p.Cys972Tyr    LB/B     rs1800433      -\n"
     ]
    }
   ],
   "source": [
    "!head -50 ../dataset/humsavar/humsavar.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "horizontal-baseline",
   "metadata": {},
   "source": [
    "## Convertion to dataframe\n",
    "The humsavar dataset is contained in a text file. For easier processing I convert it to a pandas dataframe."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "restricted-supervision",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "A1BG      P04217     VAR_018369  p.His52Arg     LB/B     rs893184       -\n",
      "A1BG      P04217     VAR_018370  p.His395Arg    LB/B     rs2241788      -\n",
      "A1CF      Q9NQ94     VAR_052201  p.Val555Met    LB/B     rs9073         -\n",
      "A1CF      Q9NQ94     VAR_059821  p.Ala558Ser    LB/B     rs11817448     -\n",
      "A2M       P01023     VAR_000012  p.Arg704His    LB/B     rs1800434      -\n",
      "A2M       P01023     VAR_000013  p.Cys972Tyr    LB/B     rs1800433      -\n",
      "A2M       P01023     VAR_000014  p.Ile1000Val   LB/B     rs669          -\n",
      "A2M       P01023     VAR_026820  p.Asn639Asp    LB/B     rs226405       -\n",
      "A2M       P01023     VAR_026821  p.Leu815Gln    LB/B     rs3180392      -\n",
      "A2ML1     A8K2U0     VAR_055463  p.Gly207Arg    LB/B     rs11047499     -\n"
     ]
    }
   ],
   "source": [
    "!cat ../dataset/humsavar/humsavar.txt|tail +45|head -n-5|tee ../processing/humsavar/humsavar_noheader.fwf|head"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "adequate-devil",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "df = pd.read_fwf('../processing/humsavar/humsavar_noheader.fwf',\n",
    "            na_values='-',\n",
    "            names=['gene_name', 'swissprot_id', 'ft_id', 'mutation', 'variant_category', 'db_snp', 'description'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "chubby-florida",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>swissprot_id</th>\n",
       "      <th>ft_id</th>\n",
       "      <th>mutation</th>\n",
       "      <th>variant_category</th>\n",
       "      <th>db_snp</th>\n",
       "      <th>description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A1BG</td>\n",
       "      <td>P04217</td>\n",
       "      <td>VAR_018369</td>\n",
       "      <td>p.His52Arg</td>\n",
       "      <td>LB/B</td>\n",
       "      <td>rs893184</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A1BG</td>\n",
       "      <td>P04217</td>\n",
       "      <td>VAR_018370</td>\n",
       "      <td>p.His395Arg</td>\n",
       "      <td>LB/B</td>\n",
       "      <td>rs2241788</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>Q9NQ94</td>\n",
       "      <td>VAR_052201</td>\n",
       "      <td>p.Val555Met</td>\n",
       "      <td>LB/B</td>\n",
       "      <td>rs9073</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>Q9NQ94</td>\n",
       "      <td>VAR_059821</td>\n",
       "      <td>p.Ala558Ser</td>\n",
       "      <td>LB/B</td>\n",
       "      <td>rs11817448</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A2M</td>\n",
       "      <td>P01023</td>\n",
       "      <td>VAR_000012</td>\n",
       "      <td>p.Arg704His</td>\n",
       "      <td>LB/B</td>\n",
       "      <td>rs1800434</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79371</th>\n",
       "      <td>NaN</td>\n",
       "      <td>Q8N402</td>\n",
       "      <td>VAR_042675</td>\n",
       "      <td>p.Thr131Pro</td>\n",
       "      <td>LB/B</td>\n",
       "      <td>rs6519442</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79372</th>\n",
       "      <td>NaN</td>\n",
       "      <td>Q96M66</td>\n",
       "      <td>VAR_039178</td>\n",
       "      <td>p.Arg37His</td>\n",
       "      <td>LB/B</td>\n",
       "      <td>rs350229</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79373</th>\n",
       "      <td>NaN</td>\n",
       "      <td>Q96M66</td>\n",
       "      <td>VAR_039179</td>\n",
       "      <td>p.Arg171Ser</td>\n",
       "      <td>LB/B</td>\n",
       "      <td>rs11648228</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79374</th>\n",
       "      <td>NaN</td>\n",
       "      <td>Q9N2K0</td>\n",
       "      <td>VAR_017799</td>\n",
       "      <td>p.Val81Leu</td>\n",
       "      <td>LB/B</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79375</th>\n",
       "      <td>NaN</td>\n",
       "      <td>Q9N2K0</td>\n",
       "      <td>VAR_017800</td>\n",
       "      <td>p.Phe150Leu</td>\n",
       "      <td>LB/B</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>79376 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      gene_name swissprot_id       ft_id     mutation variant_category  \\\n",
       "0          A1BG       P04217  VAR_018369   p.His52Arg             LB/B   \n",
       "1          A1BG       P04217  VAR_018370  p.His395Arg             LB/B   \n",
       "2          A1CF       Q9NQ94  VAR_052201  p.Val555Met             LB/B   \n",
       "3          A1CF       Q9NQ94  VAR_059821  p.Ala558Ser             LB/B   \n",
       "4           A2M       P01023  VAR_000012  p.Arg704His             LB/B   \n",
       "...         ...          ...         ...          ...              ...   \n",
       "79371       NaN       Q8N402  VAR_042675  p.Thr131Pro             LB/B   \n",
       "79372       NaN       Q96M66  VAR_039178   p.Arg37His             LB/B   \n",
       "79373       NaN       Q96M66  VAR_039179  p.Arg171Ser             LB/B   \n",
       "79374       NaN       Q9N2K0  VAR_017799   p.Val81Leu             LB/B   \n",
       "79375       NaN       Q9N2K0  VAR_017800  p.Phe150Leu             LB/B   \n",
       "\n",
       "           db_snp description  \n",
       "0        rs893184         NaN  \n",
       "1       rs2241788         NaN  \n",
       "2          rs9073         NaN  \n",
       "3      rs11817448         NaN  \n",
       "4       rs1800434         NaN  \n",
       "...           ...         ...  \n",
       "79371   rs6519442         NaN  \n",
       "79372    rs350229         NaN  \n",
       "79373  rs11648228         NaN  \n",
       "79374         NaN         NaN  \n",
       "79375         NaN         NaN  \n",
       "\n",
       "[79376 rows x 7 columns]"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "skilled-frost",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
