{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "related-trinity",
   "metadata": {},
   "source": [
    "# DMS training data creation\n",
    "\n",
    "Here I compile my own csv dataset to be used for training with different methods. I use DMS data.\n",
    "\n",
    "TLDR: Just run the entire notebook to recreate the csv dataset\n",
    "\n",
    "## DMS studies\n",
    "\n",
    "Many of the proteins studied are single domains, and I think it is useful to refer to single domains also in the input used.\n",
    "Each training protein is identified by the uniprot identifier and the position considered.\n",
    "The selection of the positions to be included is done manually after reading the original publications.\n",
    "I report here only the proteins for which I do not use the entire uniprot sequence.\n",
    "\n",
    "### Beta lactamase - P62593\n",
    "\n",
    "They used the full protein and selected on survival on antibiotic. The mutated position extend 1 after the uniprot sequence (second last is coherent with the last uniprot position).\n",
    "\n",
    "- Region: full\n",
    "- Paper: Firnberg2014\n",
    "- DMS_id: beta-lactamase\n",
    "\n",
    "### WW domain (Yap65) - P46937\n",
    "\n",
    "They used the first WW domain of human Yap65. The positions represented in the DMS experiment are 170-203 while the Interpro WW domain is in position 171-204. To be sure to include all the relevant positions I select the Interpro superfamily which covers 167-207. They selected by binding the peptide GTPPPPYTVG using phage display. The WW domain in general binds proline rich peptides.\n",
    "\n",
    "- Region: 167-207\n",
    "- Paper: Fowler2010\n",
    "- DMS_id: WW_domain\n",
    "\n",
    "### PSD95 (pdz3 ddomain) - P31016\n",
    "\n",
    "They only mutated the third PDZ domain (pdz3), position 311-393. The Interpro domain is 313-394. The interpro superfamily is 302-426.\n",
    "I use the superfamily to be permissive.\n",
    "Binds the peptide TKNYKQTSV-COOH, from the cysteine-rich interactor of PDZ (CRIPT). They selected by binding this peptide. They measured ligand binding via GFP and FACS.\n",
    "\n",
    "- Region: 302-426\n",
    "- Paper: McLaughlin2012\n",
    "- DMS_id: PSD95pdz3\n",
    "\n",
    "### Aminoglycoside kinase - P00552\n",
    "\n",
    "They used the full protein (I checked the positions, form the first to the last). This protein confers resistance to aminoglicoside antibiotics. It was isolated in K. pneumoniae and has been traansferred in many bacteria. It is coded by the transposon Tn5. They repeated the experiment with different selection strategies (antibiotic type and concentration).\n",
    "I suppose that 1:2 inthe gray2018 dms_id name refers to the concentration of kanamycin used (1/2 of the WT MIC).\n",
    "\n",
    "- Region: full\n",
    "- Paper: Melnikov2014\n",
    "- DMS_id: kka2_1:2\n",
    "\n",
    "### Ubiquitin yeast - P0CG63\n",
    "\n",
    "- Region: 1-76\n",
    "- Paper: Roscoe2013, Roscoe2014\n",
    "- DMS_id: Ubiquitin, E1_Ubiquitin\n",
    "\n",
    "Ubiquitin is code as a poly-ubiquitin gene, but all the mutations in the paper refer to a single ubiquitin repeat. I use a single repeat as input sequence. In Ubiquitin they did a growth assay with a plasmid in a yeast strain with a conditional ubiquitin knockout. In E1_Ubiquitin they used a yeast display approach and detected with FACS.\n",
    "\n",
    "### Hsp90 yeast - P02829\n",
    "\n",
    "Only the ATPase domain is considered in the study. They did a growth assay.\n",
    "\n",
    "- Region: 2-231\n",
    "- Paper: Mishra2016\n",
    "- DMS_id: hsp90\n",
    "\n",
    "The trimmed sequence generates better alignments for ev couplings since the full sequence tends to recruit sequences that are homologs only for a different, unrelated domain.\n",
    "\n",
    "### IgG binding protein Streptococcus - P06654\n",
    "\n",
    "Only the binding domain 1 (GB1) is considered. They selected according to the binding to the IgG Fc portion. They used an mRNA display approach.\n",
    "\n",
    "- Region: 226-283\n",
    "- Paper: Olson2014\n",
    "- DMS_id: gb1\n",
    "\n",
    "### Pab1 (Poly-A binding protein) - P04147\n",
    "\n",
    "They mutated the RRM (Rna Recognition Motif) domain, which binds the poly-A tail of mRNA. This specific RRM domain of the protein (domain 2 of 4) also binds to eIF4G at the cap of the mRNA. The cited paper (Melamed2015) is not actually the primary study, but a re-analisys of the data done by the same group (Melamed2013). They did a growth assay. Mutations are in position 126-200, and the Interpro RRM2 domain in position 126-203. I use the interpro domain.\n",
    "\n",
    "- Region: 126-203\n",
    "- Paper: Melamed2013\n",
    "- DMS_id: Pab1\n",
    "\n",
    "## Trimming to the mutated domain\n",
    "\n",
    "Trimming seems to be useful for some proteins where the untrimmed MSA is very wrong (hsp90, ubiquitin) and indifferent or deleterious in other cases. For trrosetta, trimming seems to always improve the contact maps. I will trim always the inputs for trrosetta, and only for hsp90 and ubiquitin for the other features.\n",
    "\n",
    "## Creation of basenames for the features\n",
    "\n",
    "All the features are extracted from the sequence, so it makes sence to refer to all the features derived from a given sequence with a unique ID. I call this `feature_basename`. More than 1 dms experiment can refer to the same basename if they correspond to mutations in the same wt sequence. I report the mapping of basenames and mutations in the dataset as appropriate. The basenames have the format `<uniprotid>_<start>-<end>`. This snippet generates the basenames from start, end and uniprot id and places them in a new column in the dataframe."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "typical-raising",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "\n",
    "def recreate_basenames(df_path):\n",
    "    df = pd.read_csv(df_path)\n",
    "    df[\"feature_basename\"] = (\n",
    "        df.uniprot_id\n",
    "        + \"_\"\n",
    "        + df.uniprot_first.astype(str)\n",
    "        + \"-\"\n",
    "        + df.uniprot_last.astype(str)\n",
    "    )\n",
    "    df.loc[\n",
    "        ((df.uniprot_last == -1) & (df.uniprot_first == 1)), [\"feature_basename\"]\n",
    "    ] = df[((df.uniprot_last == -1) & (df.uniprot_first == 1))].uniprot_id\n",
    "    df[\"feature_basename_trrosetta\"] = (\n",
    "        df.uniprot_id\n",
    "        + \"_\"\n",
    "        + df.mutated_domain_uniprot_first.astype(str)\n",
    "        + \"-\"\n",
    "        + df.mutated_domain_uniprot_last.astype(str)\n",
    "    )\n",
    "    df.loc[\n",
    "        (\n",
    "            (df.mutated_domain_uniprot_last == -1)\n",
    "            & (df.mutated_domain_uniprot_first == 1)\n",
    "        ),\n",
    "        [\"feature_basename_trrosetta\"],\n",
    "    ] = df[\n",
    "        (\n",
    "            (df.mutated_domain_uniprot_last == -1)\n",
    "            & (df.mutated_domain_uniprot_first == 1)\n",
    "        )\n",
    "    ].uniprot_id\n",
    "    df.to_csv(df_path, index=False)\n",
    "\n",
    "\n",
    "recreate_basenames(\"~/master_thesis_work/dataset/dms_datasets.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "rapid-punishment",
   "metadata": {},
   "source": [
    "## Substitution of the kka2_1:2 dataset with data from the original publication\n",
    "\n",
    "After some exploratory analysis it was observed that the version of the dataset kka2_1:2 included in gray2018 contained some duplicate entries.\n",
    "To avoid possible biases, it is replaced with the original data from the publication for that dataset, melnikov2014.\n",
    "Since that dataset includes many different experiments (different selection strategies), I selected only one of them.\n",
    "This is `KKA2_S3_Kan12_L1` and I already computed it in tidy format using the notebook `kka2_1:2_inspection.Rmd`.\n",
    "Here I replace the fitness scores for this dataset in my training set."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "rocky-victoria",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# extraction of the data that I need from the gray2018 summary\n",
    "gray2018_df = pd.read_csv(\"../../dataset/gray2018/dmsTraining_2017-02-20.csv\")\n",
    "\n",
    "# the data from the original publication\n",
    "melnikov2014_df = pd.read_csv(\"../../dataset/kka2_1:2_scores_from_melnikov2014.csv\")\n",
    "\n",
    "# add the same fields already in gray2018_df\n",
    "melnikov2014_df[\"dms_id\"] = \"kka2_1:2\"\n",
    "melnikov2014_df[\"protein\"] = \"Kka2\"\n",
    "melnikov2014_df[\"uniprot_id\"] = \"P00552\"\n",
    "\n",
    "# remove the original kka2_1:2 data and replace with the new one\n",
    "gray2018_df_kka2_replaced = pd.concat(\n",
    "    [gray2018_df[gray2018_df.dms_id != \"kka2_1:2\"], melnikov2014_df]\n",
    ").reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "interior-operations",
   "metadata": {},
   "source": [
    "## Extraction of functional scores\n",
    "\n",
    "I extract the functional scores from the Gray2018 dataset and merge the dataframe of basenames with it so to have each mutation paired to a basename."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "adolescent-accommodation",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "# correct wrong entries\n",
    "gray2018_df_kka2_replaced.loc[\n",
    "    (gray2018_df_kka2_replaced.dms_id == \"Ubiquitin\"), \"uniprot_id\"\n",
    "] = \"P0CG63\"\n",
    "cols_gray2018_to_consider = [\n",
    "    \"protein\",\n",
    "    \"dms_id\",\n",
    "    \"uniprot_id\",\n",
    "    \"position\",\n",
    "    \"aa1\",\n",
    "    \"aa2\",\n",
    "    \"reported_fitness\",\n",
    "]\n",
    "gray2018_df_kka2_replaced = gray2018_df_kka2_replaced[cols_gray2018_to_consider]\n",
    "\n",
    "# uniprot sequences ranges to consider (both extremes included) and basenames for each study\n",
    "dms_datasets_df = pd.read_csv(\"../../dataset/dms_datasets.csv\")\n",
    "\n",
    "df_raw = pd.merge(gray2018_df_kka2_replaced, dms_datasets_df)\n",
    "\n",
    "# compute the correct index in the feature vectors\n",
    "# not that this is the 0-indexed value since position is 1-indexed and also uniprot_first\n",
    "# by subtracting uniprot_first I am subtracting 1 more than neeeded for getting the 1-indexed position\n",
    "df_raw[\"feature_index\"] = df_raw.position - df_raw.uniprot_first\n",
    "# this is the position 1-indexed\n",
    "df_raw[\"feature_position\"] = df_raw.feature_index + 1\n",
    "\n",
    "# check that my mapping is correct (I am not removing needed positions)\n",
    "assert np.all((df_raw.position <= df_raw.uniprot_last) | (df_raw.uniprot_last == -1))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "extended-converter",
   "metadata": {},
   "source": [
    "## Removal of problematic entries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "casual-hurricane",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "hsp90 4417 4231\n",
      "beta-lactamase 5436 5397\n",
      "WW_domain 377 373\n",
      "PSD95pdz3 1577 1577\n",
      "gb1 1045 1026\n",
      "E1_Ubiquitin 1198 1142\n",
      "kka2_1:2 5280 5280\n",
      "Pab1 1276 1220\n",
      "Ubiquitin 1403 1267\n",
      "TOTAL 22009 21513\n",
      "Number of mutations removed: 496\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "\n",
    "# I remove positions with unknown mutations\n",
    "df = df_raw[df_raw.aa2 != \"X\"]\n",
    "\n",
    "# I remove positions where the aa1 does not coincide with the one from uniprot\n",
    "uniprot_seq_df_all_proteins = pd.DataFrame()\n",
    "for basename in set(df.feature_basename):\n",
    "    uniprot_seq = np.array(\n",
    "        SeqIO.read(\n",
    "            \"../../processing/uniprot_sequences/\" + basename + \".fasta\", \"fasta\"\n",
    "        ).seq\n",
    "    )\n",
    "    uniprot_seq_df = pd.DataFrame(\n",
    "        {\n",
    "            \"aa1\": uniprot_seq,\n",
    "            \"feature_position\": range(1, uniprot_seq.shape[0] + 1),\n",
    "            \"feature_basename\": basename,\n",
    "        }\n",
    "    )\n",
    "    uniprot_seq_df_all_proteins = pd.concat(\n",
    "        [uniprot_seq_df_all_proteins, uniprot_seq_df]\n",
    "    )\n",
    "\n",
    "# I want the intersection between the 2 df (where aa1 is the same)\n",
    "df = pd.merge(df, uniprot_seq_df_all_proteins)\n",
    "\n",
    "for study in set(df.dms_id):\n",
    "    print(study, len(df_raw[df_raw.dms_id == study]), len(df[df.dms_id == study]))\n",
    "print(\"TOTAL\", len(df_raw), len(df))\n",
    "print(\"Number of mutations removed:\", len(df_raw) - len(df))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "alternative-annual",
   "metadata": {},
   "source": [
    "## Addition of normalised scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "sorted-headset",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.preprocessing import QuantileTransformer\n",
    "\n",
    "for study in set(df.dms_id):\n",
    "    y_curr = df[df.dms_id == study].reported_fitness.to_numpy().reshape(-1, 1)\n",
    "    scaler = QuantileTransformer(n_quantiles=100)\n",
    "    scaler.fit(y_curr)\n",
    "    df.loc[(df.dms_id == study), [\"reported_fitness_quantile\"]] = scaler.transform(\n",
    "        y_curr\n",
    "    ) - scaler.transform([[0]])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "differential-mumbai",
   "metadata": {},
   "source": [
    "## Adding features\n",
    "\n",
    "The features are loaded from a variety of inputs and always contain the basename in the filename. Here I add them to the dataframe."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "connected-extent",
   "metadata": {},
   "outputs": [],
   "source": [
    "import glob\n",
    "import os\n",
    "\n",
    "import joblib\n",
    "import networkx as nx\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from sklearn.preprocessing import (\n",
    "    MaxAbsScaler,\n",
    "    MinMaxScaler,\n",
    "    QuantileTransformer,\n",
    "    RobustScaler,\n",
    "    StandardScaler,\n",
    ")\n",
    "\n",
    "# load the ev mutation output\n",
    "ev_df_all_proteins = pd.DataFrame()\n",
    "for basename in set(df.feature_basename):\n",
    "    ev_df = pd.read_csv(\n",
    "        \"../../processing/ev_couplings/\" + basename + \"_single_mutant_matrix.csv\"\n",
    "    )\n",
    "    ev_df_simple = pd.DataFrame(\n",
    "        {\n",
    "            \"feature_position\": ev_df.pos,\n",
    "            \"aa1\": ev_df.wt,\n",
    "            \"aa2\": ev_df.subs,\n",
    "            \"ev_frequency\": ev_df.frequency,\n",
    "            \"ev_conservation\": ev_df.column_conservation,\n",
    "            \"ev_independent\": ev_df.prediction_independent,\n",
    "            \"ev_epistatic\": ev_df.prediction_epistatic,\n",
    "        }\n",
    "    )\n",
    "    ev_df_simple[\"feature_basename\"] = basename\n",
    "    ev_df_all_proteins = pd.concat([ev_df_all_proteins, ev_df_simple])\n",
    "\n",
    "df = pd.merge(df, ev_df_all_proteins, how=\"left\")\n",
    "\n",
    "# load the netsurf output\n",
    "netsurf_df_all_proteins = pd.DataFrame()\n",
    "for basename in set(df.feature_basename):\n",
    "    netsurf_df = pd.read_csv(\"../../processing/netsurfp2/\" + basename + \"_netsurf.csv\")\n",
    "    netsurf_df_simple = pd.DataFrame(\n",
    "        {\n",
    "            \"feature_position\": netsurf_df.n,\n",
    "            \"aa1\": netsurf_df.seq,\n",
    "        }\n",
    "    )\n",
    "    netsurf_feature_names = [\n",
    "        \"rsa\",\n",
    "        \"asa\",\n",
    "        \"q3\",\n",
    "        \"p[q3_H]\",\n",
    "        \"p[q3_E]\",\n",
    "        \"p[q3_C]\",\n",
    "        \"q8\",\n",
    "        \"p[q8_G]\",\n",
    "        \"p[q8_H]\",\n",
    "        \"p[q8_I]\",\n",
    "        \"p[q8_B]\",\n",
    "        \"p[q8_E]\",\n",
    "        \"p[q8_S]\",\n",
    "        \"p[q8_T]\",\n",
    "        \"p[q8_C]\",\n",
    "        \"phi\",\n",
    "        \"psi\",\n",
    "        \"disorder\",\n",
    "    ]\n",
    "    for label in netsurf_feature_names:\n",
    "        new_label = \"netsurf_\" + label\n",
    "        netsurf_df_simple[new_label] = netsurf_df[label]\n",
    "    netsurf_df_simple[\"feature_basename\"] = basename\n",
    "    netsurf_df_all_proteins = pd.concat([netsurf_df_all_proteins, netsurf_df_simple])\n",
    "\n",
    "df = pd.merge(df, netsurf_df_all_proteins, how=\"left\")\n",
    "\n",
    "# load the hmmer pssm\n",
    "hmm_pssm_df_all_proteins = pd.DataFrame()\n",
    "for basename in set(df.feature_basename):\n",
    "    hmm_pssm_vec = joblib.load(\n",
    "        \"../../processing/hmmer/\" + basename + \".hmm_pssm.joblib.xz\"\n",
    "    )[\"pssm\"]\n",
    "    hmm_pssm_colnames = joblib.load(\n",
    "        \"../../processing/hmmer/\" + basename + \".hmm_pssm.joblib.xz\"\n",
    "    )[\"colnames\"]\n",
    "    hmm_pssm_df_simple = pd.DataFrame(\n",
    "        {\n",
    "            # the position is 1-indexed and the range 0-indexed\n",
    "            # entries in the profile are in the correct order\n",
    "            \"feature_position\": np.array(range(len(hmm_pssm_vec)))\n",
    "            + 1,\n",
    "        }\n",
    "    )\n",
    "    for i, residue in enumerate(hmm_pssm_colnames):\n",
    "        hmm_pssm_df_simple[\"hmm_pssm_\" + residue] = hmm_pssm_vec[:, i]\n",
    "    hmm_pssm_df_simple[\"feature_basename\"] = basename\n",
    "    hmm_pssm_df_all_proteins = pd.concat([hmm_pssm_df_all_proteins, hmm_pssm_df_simple])\n",
    "\n",
    "df = pd.merge(df, hmm_pssm_df_all_proteins, how=\"left\")\n",
    "\n",
    "# compute the likelyhood of the mutation from the pssm\n",
    "df[\"hmm_pssm_aa1_likelyhood\"] = [row[\"hmm_pssm_\" + row.aa1] for _, row in df.iterrows()]\n",
    "df[\"hmm_pssm_aa2_likelyhood\"] = [row[\"hmm_pssm_\" + row.aa2] for _, row in df.iterrows()]\n",
    "df[\"hmm_pssm_delta_likelyhood\"] = (\n",
    "    df.hmm_pssm_aa2_likelyhood - df.hmm_pssm_aa1_likelyhood\n",
    ")\n",
    "\n",
    "# add graph metrics from trRosetta\n",
    "contact_treshold = 8\n",
    "tr_rosetta_graph_df_all_proteins = pd.DataFrame()\n",
    "for trrosetta_basename in set(df.feature_basename_trrosetta):\n",
    "    trrosetta_uniprot_firsts = set(\n",
    "        df[\n",
    "            df.feature_basename_trrosetta == trrosetta_basename\n",
    "        ].mutated_domain_uniprot_first\n",
    "    )\n",
    "    assert len(trrosetta_uniprot_firsts) == 1\n",
    "    trrosetta_uniprot_first = trrosetta_uniprot_firsts.pop()\n",
    "    tr_rosetta_distances = joblib.load(\n",
    "        \"../../processing/tr_rosetta/{}_trRosetta_distance_mat.joblib.xz\".format(\n",
    "            trrosetta_basename\n",
    "        )\n",
    "    )\n",
    "    G = nx.Graph()\n",
    "    for i, row in enumerate(tr_rosetta_distances):\n",
    "        G.add_node(i)\n",
    "        for j, el in enumerate(row):\n",
    "            if i >= j and el < contact_treshold:\n",
    "                G.add_edge(i, j)\n",
    "    assert G.number_of_nodes() == tr_rosetta_distances.shape[0]\n",
    "    graph_df = pd.DataFrame()\n",
    "    graph_df[\"tr_rosetta_feature_index\"] = range(tr_rosetta_distances.shape[0])\n",
    "    graph_df[\"tr_rosetta_graph_closeness_centrality\"] = graph_df[\n",
    "        \"tr_rosetta_feature_index\"\n",
    "    ].map(nx.closeness_centrality(G))\n",
    "    graph_df[\"tr_rosetta_graph_betweenness_centrality\"] = graph_df[\n",
    "        \"tr_rosetta_feature_index\"\n",
    "    ].map(nx.betweenness_centrality(G, normalized=True))\n",
    "    graph_df[\"tr_rosetta_graph_degree_centrality\"] = graph_df[\n",
    "        \"tr_rosetta_feature_index\"\n",
    "    ].map(nx.degree_centrality(G))\n",
    "    graph_df[\"tr_rosetta_graph_load_centrality\"] = graph_df[\n",
    "        \"tr_rosetta_feature_index\"\n",
    "    ].map(nx.load_centrality(G, normalized=True))\n",
    "    graph_df[\"tr_rosetta_graph_harmonic_centrality\"] = (\n",
    "        graph_df[\"tr_rosetta_feature_index\"].map(nx.harmonic_centrality(G)) / 100\n",
    "    )\n",
    "    graph_df[\"tr_rosetta_graph_clustering\"] = graph_df[\"tr_rosetta_feature_index\"].map(\n",
    "        nx.clustering(G)\n",
    "    )\n",
    "    graph_df[\"feature_basename_trrosetta\"] = trrosetta_basename\n",
    "    # this is a position and not an index since trtrrosetta_uniprot_first is 1-indexed\n",
    "    # It is position and not feature_position since it refers to the full uniprot seq\n",
    "    graph_df[\"position\"] = (\n",
    "        trrosetta_uniprot_first + graph_df[\"tr_rosetta_feature_index\"]\n",
    "    )\n",
    "    tr_rosetta_graph_df_all_proteins = pd.concat(\n",
    "        [tr_rosetta_graph_df_all_proteins, graph_df]\n",
    "    )\n",
    "\n",
    "df = pd.merge(df, tr_rosetta_graph_df_all_proteins, how=\"left\")\n",
    "\n",
    "# add dssp output (not a feature, but for comparison)\n",
    "dssp_df_all_proteins = pd.DataFrame()\n",
    "for uniprot_id in set(df.uniprot_id):\n",
    "    file_glob = glob.glob(\n",
    "        \"../../processing/structures/dssp_mapped/{}_mapped_*_*.uniprot_dssp.joblib.xz\".format(\n",
    "            uniprot_id\n",
    "        )\n",
    "    )\n",
    "    assert len(file_glob) == 1\n",
    "    filename = file_glob[0]\n",
    "    dssp_df_raw = joblib.load(filename)\n",
    "    dssp_df_simple = pd.DataFrame()\n",
    "    dssp_df_simple[\"position\"] = dssp_df_raw[\"uniprot_res_pos\"]\n",
    "    dssp_df_simple[\"uniprot_id\"] = uniprot_id\n",
    "    dssp_df_simple[\"aa1\"] = dssp_df_raw[\"residue_type\"]\n",
    "    dssp_df_simple[\"dssp_sec_struct\"] = dssp_df_raw[\"secondary_structure\"]\n",
    "    dssp_df_simple[\"dssp_rsa\"] = dssp_df_raw[\"relative_solvent_accessibility\"]\n",
    "    dssp_df_simple[\"dssp_asa\"] = dssp_df_raw[\"solvent_accessibility\"]\n",
    "    dssp_df_simple[\"dssp_phi\"] = dssp_df_raw[\"phi\"]\n",
    "    dssp_df_simple[\"dssp_psi\"] = dssp_df_raw[\"psi\"]\n",
    "    if uniprot_id == \"P06654\":\n",
    "        # the gb1 structure is not available but it is availbale a structure for the next IgG binding domain\n",
    "        dssp_df_simple[\"position\"] = dssp_df_raw[\"uniprot_res_pos\"] - 70\n",
    "    elif uniprot_id == \"P0CG63\":\n",
    "        # the structure is about the third ubiquitin repeat, the mutated positions refer to the first\n",
    "        dssp_df_simple[\"position\"] = dssp_df_raw[\"uniprot_res_pos\"] - 304\n",
    "    dssp_df_all_proteins = pd.concat([dssp_df_all_proteins, dssp_df_simple])\n",
    "\n",
    "df = pd.merge(df, dssp_df_all_proteins, how=\"left\")\n",
    "\n",
    "# sanify the feature names for xgb (cannot contain \\[, \\], or \\,)\n",
    "df = df.rename(\n",
    "    {col: col.replace(\"[\", \"_\").replace(\"]\", \"\") for col in df.columns}, axis=\"columns\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "growing-mother",
   "metadata": {},
   "source": [
    "## Export the data\n",
    "\n",
    "I save the dataframe with all the mappings, features, and labels in a csv file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "sought-connection",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein</th>\n",
       "      <th>dms_id</th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>position</th>\n",
       "      <th>aa1</th>\n",
       "      <th>aa2</th>\n",
       "      <th>reported_fitness</th>\n",
       "      <th>pdb_id</th>\n",
       "      <th>pdb_chain</th>\n",
       "      <th>author_year</th>\n",
       "      <th>...</th>\n",
       "      <th>tr_rosetta_graph_betweenness_centrality</th>\n",
       "      <th>tr_rosetta_graph_degree_centrality</th>\n",
       "      <th>tr_rosetta_graph_load_centrality</th>\n",
       "      <th>tr_rosetta_graph_harmonic_centrality</th>\n",
       "      <th>tr_rosetta_graph_clustering</th>\n",
       "      <th>dssp_sec_struct</th>\n",
       "      <th>dssp_rsa</th>\n",
       "      <th>dssp_asa</th>\n",
       "      <th>dssp_phi</th>\n",
       "      <th>dssp_psi</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TEM-1</td>\n",
       "      <td>beta-lactamase</td>\n",
       "      <td>P62593</td>\n",
       "      <td>20</td>\n",
       "      <td>P</td>\n",
       "      <td>P</td>\n",
       "      <td>0.581033</td>\n",
       "      <td>1btl</td>\n",
       "      <td>A</td>\n",
       "      <td>Firnberg2014</td>\n",
       "      <td>...</td>\n",
       "      <td>0.116044</td>\n",
       "      <td>0.017544</td>\n",
       "      <td>0.111866</td>\n",
       "      <td>0.42778</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TEM-1</td>\n",
       "      <td>beta-lactamase</td>\n",
       "      <td>P62593</td>\n",
       "      <td>20</td>\n",
       "      <td>P</td>\n",
       "      <td>Q</td>\n",
       "      <td>0.441480</td>\n",
       "      <td>1btl</td>\n",
       "      <td>A</td>\n",
       "      <td>Firnberg2014</td>\n",
       "      <td>...</td>\n",
       "      <td>0.116044</td>\n",
       "      <td>0.017544</td>\n",
       "      <td>0.111866</td>\n",
       "      <td>0.42778</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TEM-1</td>\n",
       "      <td>beta-lactamase</td>\n",
       "      <td>P62593</td>\n",
       "      <td>20</td>\n",
       "      <td>P</td>\n",
       "      <td>D</td>\n",
       "      <td>0.289750</td>\n",
       "      <td>1btl</td>\n",
       "      <td>A</td>\n",
       "      <td>Firnberg2014</td>\n",
       "      <td>...</td>\n",
       "      <td>0.116044</td>\n",
       "      <td>0.017544</td>\n",
       "      <td>0.111866</td>\n",
       "      <td>0.42778</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TEM-1</td>\n",
       "      <td>beta-lactamase</td>\n",
       "      <td>P62593</td>\n",
       "      <td>20</td>\n",
       "      <td>P</td>\n",
       "      <td>K</td>\n",
       "      <td>0.196582</td>\n",
       "      <td>1btl</td>\n",
       "      <td>A</td>\n",
       "      <td>Firnberg2014</td>\n",
       "      <td>...</td>\n",
       "      <td>0.116044</td>\n",
       "      <td>0.017544</td>\n",
       "      <td>0.111866</td>\n",
       "      <td>0.42778</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TEM-1</td>\n",
       "      <td>beta-lactamase</td>\n",
       "      <td>P62593</td>\n",
       "      <td>20</td>\n",
       "      <td>P</td>\n",
       "      <td>N</td>\n",
       "      <td>0.053725</td>\n",
       "      <td>1btl</td>\n",
       "      <td>A</td>\n",
       "      <td>Firnberg2014</td>\n",
       "      <td>...</td>\n",
       "      <td>0.116044</td>\n",
       "      <td>0.017544</td>\n",
       "      <td>0.111866</td>\n",
       "      <td>0.42778</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21508</th>\n",
       "      <td>Kka2</td>\n",
       "      <td>kka2_1:2</td>\n",
       "      <td>P00552</td>\n",
       "      <td>64</td>\n",
       "      <td>A</td>\n",
       "      <td>L</td>\n",
       "      <td>-3.278095</td>\n",
       "      <td>1nd4</td>\n",
       "      <td>A</td>\n",
       "      <td>Melnikov2014</td>\n",
       "      <td>...</td>\n",
       "      <td>0.023451</td>\n",
       "      <td>0.041825</td>\n",
       "      <td>0.025188</td>\n",
       "      <td>0.69352</td>\n",
       "      <td>0.490909</td>\n",
       "      <td>H</td>\n",
       "      <td>0.072595</td>\n",
       "      <td>8.0</td>\n",
       "      <td>-55.2</td>\n",
       "      <td>-52.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21509</th>\n",
       "      <td>Kka2</td>\n",
       "      <td>kka2_1:2</td>\n",
       "      <td>P00552</td>\n",
       "      <td>64</td>\n",
       "      <td>A</td>\n",
       "      <td>F</td>\n",
       "      <td>0.450056</td>\n",
       "      <td>1nd4</td>\n",
       "      <td>A</td>\n",
       "      <td>Melnikov2014</td>\n",
       "      <td>...</td>\n",
       "      <td>0.023451</td>\n",
       "      <td>0.041825</td>\n",
       "      <td>0.025188</td>\n",
       "      <td>0.69352</td>\n",
       "      <td>0.490909</td>\n",
       "      <td>H</td>\n",
       "      <td>0.072595</td>\n",
       "      <td>8.0</td>\n",
       "      <td>-55.2</td>\n",
       "      <td>-52.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21510</th>\n",
       "      <td>Kka2</td>\n",
       "      <td>kka2_1:2</td>\n",
       "      <td>P00552</td>\n",
       "      <td>64</td>\n",
       "      <td>A</td>\n",
       "      <td>I</td>\n",
       "      <td>-2.099644</td>\n",
       "      <td>1nd4</td>\n",
       "      <td>A</td>\n",
       "      <td>Melnikov2014</td>\n",
       "      <td>...</td>\n",
       "      <td>0.023451</td>\n",
       "      <td>0.041825</td>\n",
       "      <td>0.025188</td>\n",
       "      <td>0.69352</td>\n",
       "      <td>0.490909</td>\n",
       "      <td>H</td>\n",
       "      <td>0.072595</td>\n",
       "      <td>8.0</td>\n",
       "      <td>-55.2</td>\n",
       "      <td>-52.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21511</th>\n",
       "      <td>Kka2</td>\n",
       "      <td>kka2_1:2</td>\n",
       "      <td>P00552</td>\n",
       "      <td>64</td>\n",
       "      <td>A</td>\n",
       "      <td>Q</td>\n",
       "      <td>-0.107029</td>\n",
       "      <td>1nd4</td>\n",
       "      <td>A</td>\n",
       "      <td>Melnikov2014</td>\n",
       "      <td>...</td>\n",
       "      <td>0.023451</td>\n",
       "      <td>0.041825</td>\n",
       "      <td>0.025188</td>\n",
       "      <td>0.69352</td>\n",
       "      <td>0.490909</td>\n",
       "      <td>H</td>\n",
       "      <td>0.072595</td>\n",
       "      <td>8.0</td>\n",
       "      <td>-55.2</td>\n",
       "      <td>-52.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21512</th>\n",
       "      <td>Kka2</td>\n",
       "      <td>kka2_1:2</td>\n",
       "      <td>P00552</td>\n",
       "      <td>64</td>\n",
       "      <td>A</td>\n",
       "      <td>W</td>\n",
       "      <td>-0.936238</td>\n",
       "      <td>1nd4</td>\n",
       "      <td>A</td>\n",
       "      <td>Melnikov2014</td>\n",
       "      <td>...</td>\n",
       "      <td>0.023451</td>\n",
       "      <td>0.041825</td>\n",
       "      <td>0.025188</td>\n",
       "      <td>0.69352</td>\n",
       "      <td>0.490909</td>\n",
       "      <td>H</td>\n",
       "      <td>0.072595</td>\n",
       "      <td>8.0</td>\n",
       "      <td>-55.2</td>\n",
       "      <td>-52.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>21513 rows Ã— 78 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      protein          dms_id uniprot_id  position aa1 aa2  reported_fitness  \\\n",
       "0       TEM-1  beta-lactamase     P62593        20   P   P          0.581033   \n",
       "1       TEM-1  beta-lactamase     P62593        20   P   Q          0.441480   \n",
       "2       TEM-1  beta-lactamase     P62593        20   P   D          0.289750   \n",
       "3       TEM-1  beta-lactamase     P62593        20   P   K          0.196582   \n",
       "4       TEM-1  beta-lactamase     P62593        20   P   N          0.053725   \n",
       "...       ...             ...        ...       ...  ..  ..               ...   \n",
       "21508    Kka2        kka2_1:2     P00552        64   A   L         -3.278095   \n",
       "21509    Kka2        kka2_1:2     P00552        64   A   F          0.450056   \n",
       "21510    Kka2        kka2_1:2     P00552        64   A   I         -2.099644   \n",
       "21511    Kka2        kka2_1:2     P00552        64   A   Q         -0.107029   \n",
       "21512    Kka2        kka2_1:2     P00552        64   A   W         -0.936238   \n",
       "\n",
       "      pdb_id pdb_chain   author_year  ...  \\\n",
       "0       1btl         A  Firnberg2014  ...   \n",
       "1       1btl         A  Firnberg2014  ...   \n",
       "2       1btl         A  Firnberg2014  ...   \n",
       "3       1btl         A  Firnberg2014  ...   \n",
       "4       1btl         A  Firnberg2014  ...   \n",
       "...      ...       ...           ...  ...   \n",
       "21508   1nd4         A  Melnikov2014  ...   \n",
       "21509   1nd4         A  Melnikov2014  ...   \n",
       "21510   1nd4         A  Melnikov2014  ...   \n",
       "21511   1nd4         A  Melnikov2014  ...   \n",
       "21512   1nd4         A  Melnikov2014  ...   \n",
       "\n",
       "       tr_rosetta_graph_betweenness_centrality  \\\n",
       "0                                     0.116044   \n",
       "1                                     0.116044   \n",
       "2                                     0.116044   \n",
       "3                                     0.116044   \n",
       "4                                     0.116044   \n",
       "...                                        ...   \n",
       "21508                                 0.023451   \n",
       "21509                                 0.023451   \n",
       "21510                                 0.023451   \n",
       "21511                                 0.023451   \n",
       "21512                                 0.023451   \n",
       "\n",
       "       tr_rosetta_graph_degree_centrality  tr_rosetta_graph_load_centrality  \\\n",
       "0                                0.017544                          0.111866   \n",
       "1                                0.017544                          0.111866   \n",
       "2                                0.017544                          0.111866   \n",
       "3                                0.017544                          0.111866   \n",
       "4                                0.017544                          0.111866   \n",
       "...                                   ...                               ...   \n",
       "21508                            0.041825                          0.025188   \n",
       "21509                            0.041825                          0.025188   \n",
       "21510                            0.041825                          0.025188   \n",
       "21511                            0.041825                          0.025188   \n",
       "21512                            0.041825                          0.025188   \n",
       "\n",
       "       tr_rosetta_graph_harmonic_centrality tr_rosetta_graph_clustering  \\\n",
       "0                                   0.42778                    0.500000   \n",
       "1                                   0.42778                    0.500000   \n",
       "2                                   0.42778                    0.500000   \n",
       "3                                   0.42778                    0.500000   \n",
       "4                                   0.42778                    0.500000   \n",
       "...                                     ...                         ...   \n",
       "21508                               0.69352                    0.490909   \n",
       "21509                               0.69352                    0.490909   \n",
       "21510                               0.69352                    0.490909   \n",
       "21511                               0.69352                    0.490909   \n",
       "21512                               0.69352                    0.490909   \n",
       "\n",
       "      dssp_sec_struct  dssp_rsa dssp_asa  dssp_phi  dssp_psi  \n",
       "0                 NaN       NaN      NaN       NaN       NaN  \n",
       "1                 NaN       NaN      NaN       NaN       NaN  \n",
       "2                 NaN       NaN      NaN       NaN       NaN  \n",
       "3                 NaN       NaN      NaN       NaN       NaN  \n",
       "4                 NaN       NaN      NaN       NaN       NaN  \n",
       "...               ...       ...      ...       ...       ...  \n",
       "21508               H  0.072595      8.0     -55.2     -52.1  \n",
       "21509               H  0.072595      8.0     -55.2     -52.1  \n",
       "21510               H  0.072595      8.0     -55.2     -52.1  \n",
       "21511               H  0.072595      8.0     -55.2     -52.1  \n",
       "21512               H  0.072595      8.0     -55.2     -52.1  \n",
       "\n",
       "[21513 rows x 78 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.to_csv(\"../../dataset/dms_training.csv\", index=False)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "macro-lambda",
   "metadata": {},
   "outputs": [],
   "source": [
    "## a"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "occupational-matter",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein</th>\n",
       "      <th>dms_id</th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>position</th>\n",
       "      <th>aa1</th>\n",
       "      <th>aa2</th>\n",
       "      <th>reported_fitness</th>\n",
       "      <th>pdb_id</th>\n",
       "      <th>pdb_chain</th>\n",
       "      <th>author_year</th>\n",
       "      <th>...</th>\n",
       "      <th>tr_rosetta_graph_betweenness_centrality</th>\n",
       "      <th>tr_rosetta_graph_degree_centrality</th>\n",
       "      <th>tr_rosetta_graph_load_centrality</th>\n",
       "      <th>tr_rosetta_graph_harmonic_centrality</th>\n",
       "      <th>tr_rosetta_graph_clustering</th>\n",
       "      <th>dssp_sec_struct</th>\n",
       "      <th>dssp_rsa</th>\n",
       "      <th>dssp_asa</th>\n",
       "      <th>dssp_phi</th>\n",
       "      <th>dssp_psi</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TEM-1</td>\n",
       "      <td>beta-lactamase</td>\n",
       "      <td>P62593</td>\n",
       "      <td>20</td>\n",
       "      <td>P</td>\n",
       "      <td>P</td>\n",
       "      <td>0.581033</td>\n",
       "      <td>1btl</td>\n",
       "      <td>A</td>\n",
       "      <td>Firnberg2014</td>\n",
       "      <td>...</td>\n",
       "      <td>0.116044</td>\n",
       "      <td>0.017544</td>\n",
       "      <td>0.111866</td>\n",
       "      <td>0.42778</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TEM-1</td>\n",
       "      <td>beta-lactamase</td>\n",
       "      <td>P62593</td>\n",
       "      <td>20</td>\n",
       "      <td>P</td>\n",
       "      <td>Q</td>\n",
       "      <td>0.441480</td>\n",
       "      <td>1btl</td>\n",
       "      <td>A</td>\n",
       "      <td>Firnberg2014</td>\n",
       "      <td>...</td>\n",
       "      <td>0.116044</td>\n",
       "      <td>0.017544</td>\n",
       "      <td>0.111866</td>\n",
       "      <td>0.42778</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TEM-1</td>\n",
       "      <td>beta-lactamase</td>\n",
       "      <td>P62593</td>\n",
       "      <td>20</td>\n",
       "      <td>P</td>\n",
       "      <td>D</td>\n",
       "      <td>0.289750</td>\n",
       "      <td>1btl</td>\n",
       "      <td>A</td>\n",
       "      <td>Firnberg2014</td>\n",
       "      <td>...</td>\n",
       "      <td>0.116044</td>\n",
       "      <td>0.017544</td>\n",
       "      <td>0.111866</td>\n",
       "      <td>0.42778</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TEM-1</td>\n",
       "      <td>beta-lactamase</td>\n",
       "      <td>P62593</td>\n",
       "      <td>20</td>\n",
       "      <td>P</td>\n",
       "      <td>K</td>\n",
       "      <td>0.196582</td>\n",
       "      <td>1btl</td>\n",
       "      <td>A</td>\n",
       "      <td>Firnberg2014</td>\n",
       "      <td>...</td>\n",
       "      <td>0.116044</td>\n",
       "      <td>0.017544</td>\n",
       "      <td>0.111866</td>\n",
       "      <td>0.42778</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TEM-1</td>\n",
       "      <td>beta-lactamase</td>\n",
       "      <td>P62593</td>\n",
       "      <td>20</td>\n",
       "      <td>P</td>\n",
       "      <td>N</td>\n",
       "      <td>0.053725</td>\n",
       "      <td>1btl</td>\n",
       "      <td>A</td>\n",
       "      <td>Firnberg2014</td>\n",
       "      <td>...</td>\n",
       "      <td>0.116044</td>\n",
       "      <td>0.017544</td>\n",
       "      <td>0.111866</td>\n",
       "      <td>0.42778</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21508</th>\n",
       "      <td>Kka2</td>\n",
       "      <td>kka2_1:2</td>\n",
       "      <td>P00552</td>\n",
       "      <td>64</td>\n",
       "      <td>A</td>\n",
       "      <td>L</td>\n",
       "      <td>-3.278095</td>\n",
       "      <td>1nd4</td>\n",
       "      <td>A</td>\n",
       "      <td>Melnikov2014</td>\n",
       "      <td>...</td>\n",
       "      <td>0.023451</td>\n",
       "      <td>0.041825</td>\n",
       "      <td>0.025188</td>\n",
       "      <td>0.69352</td>\n",
       "      <td>0.490909</td>\n",
       "      <td>H</td>\n",
       "      <td>0.072595</td>\n",
       "      <td>8.0</td>\n",
       "      <td>-55.2</td>\n",
       "      <td>-52.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21509</th>\n",
       "      <td>Kka2</td>\n",
       "      <td>kka2_1:2</td>\n",
       "      <td>P00552</td>\n",
       "      <td>64</td>\n",
       "      <td>A</td>\n",
       "      <td>F</td>\n",
       "      <td>0.450056</td>\n",
       "      <td>1nd4</td>\n",
       "      <td>A</td>\n",
       "      <td>Melnikov2014</td>\n",
       "      <td>...</td>\n",
       "      <td>0.023451</td>\n",
       "      <td>0.041825</td>\n",
       "      <td>0.025188</td>\n",
       "      <td>0.69352</td>\n",
       "      <td>0.490909</td>\n",
       "      <td>H</td>\n",
       "      <td>0.072595</td>\n",
       "      <td>8.0</td>\n",
       "      <td>-55.2</td>\n",
       "      <td>-52.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21510</th>\n",
       "      <td>Kka2</td>\n",
       "      <td>kka2_1:2</td>\n",
       "      <td>P00552</td>\n",
       "      <td>64</td>\n",
       "      <td>A</td>\n",
       "      <td>I</td>\n",
       "      <td>-2.099644</td>\n",
       "      <td>1nd4</td>\n",
       "      <td>A</td>\n",
       "      <td>Melnikov2014</td>\n",
       "      <td>...</td>\n",
       "      <td>0.023451</td>\n",
       "      <td>0.041825</td>\n",
       "      <td>0.025188</td>\n",
       "      <td>0.69352</td>\n",
       "      <td>0.490909</td>\n",
       "      <td>H</td>\n",
       "      <td>0.072595</td>\n",
       "      <td>8.0</td>\n",
       "      <td>-55.2</td>\n",
       "      <td>-52.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21511</th>\n",
       "      <td>Kka2</td>\n",
       "      <td>kka2_1:2</td>\n",
       "      <td>P00552</td>\n",
       "      <td>64</td>\n",
       "      <td>A</td>\n",
       "      <td>Q</td>\n",
       "      <td>-0.107029</td>\n",
       "      <td>1nd4</td>\n",
       "      <td>A</td>\n",
       "      <td>Melnikov2014</td>\n",
       "      <td>...</td>\n",
       "      <td>0.023451</td>\n",
       "      <td>0.041825</td>\n",
       "      <td>0.025188</td>\n",
       "      <td>0.69352</td>\n",
       "      <td>0.490909</td>\n",
       "      <td>H</td>\n",
       "      <td>0.072595</td>\n",
       "      <td>8.0</td>\n",
       "      <td>-55.2</td>\n",
       "      <td>-52.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21512</th>\n",
       "      <td>Kka2</td>\n",
       "      <td>kka2_1:2</td>\n",
       "      <td>P00552</td>\n",
       "      <td>64</td>\n",
       "      <td>A</td>\n",
       "      <td>W</td>\n",
       "      <td>-0.936238</td>\n",
       "      <td>1nd4</td>\n",
       "      <td>A</td>\n",
       "      <td>Melnikov2014</td>\n",
       "      <td>...</td>\n",
       "      <td>0.023451</td>\n",
       "      <td>0.041825</td>\n",
       "      <td>0.025188</td>\n",
       "      <td>0.69352</td>\n",
       "      <td>0.490909</td>\n",
       "      <td>H</td>\n",
       "      <td>0.072595</td>\n",
       "      <td>8.0</td>\n",
       "      <td>-55.2</td>\n",
       "      <td>-52.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>21513 rows Ã— 78 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      protein          dms_id uniprot_id  position aa1 aa2  reported_fitness  \\\n",
       "0       TEM-1  beta-lactamase     P62593        20   P   P          0.581033   \n",
       "1       TEM-1  beta-lactamase     P62593        20   P   Q          0.441480   \n",
       "2       TEM-1  beta-lactamase     P62593        20   P   D          0.289750   \n",
       "3       TEM-1  beta-lactamase     P62593        20   P   K          0.196582   \n",
       "4       TEM-1  beta-lactamase     P62593        20   P   N          0.053725   \n",
       "...       ...             ...        ...       ...  ..  ..               ...   \n",
       "21508    Kka2        kka2_1:2     P00552        64   A   L         -3.278095   \n",
       "21509    Kka2        kka2_1:2     P00552        64   A   F          0.450056   \n",
       "21510    Kka2        kka2_1:2     P00552        64   A   I         -2.099644   \n",
       "21511    Kka2        kka2_1:2     P00552        64   A   Q         -0.107029   \n",
       "21512    Kka2        kka2_1:2     P00552        64   A   W         -0.936238   \n",
       "\n",
       "      pdb_id pdb_chain   author_year  ...  \\\n",
       "0       1btl         A  Firnberg2014  ...   \n",
       "1       1btl         A  Firnberg2014  ...   \n",
       "2       1btl         A  Firnberg2014  ...   \n",
       "3       1btl         A  Firnberg2014  ...   \n",
       "4       1btl         A  Firnberg2014  ...   \n",
       "...      ...       ...           ...  ...   \n",
       "21508   1nd4         A  Melnikov2014  ...   \n",
       "21509   1nd4         A  Melnikov2014  ...   \n",
       "21510   1nd4         A  Melnikov2014  ...   \n",
       "21511   1nd4         A  Melnikov2014  ...   \n",
       "21512   1nd4         A  Melnikov2014  ...   \n",
       "\n",
       "       tr_rosetta_graph_betweenness_centrality  \\\n",
       "0                                     0.116044   \n",
       "1                                     0.116044   \n",
       "2                                     0.116044   \n",
       "3                                     0.116044   \n",
       "4                                     0.116044   \n",
       "...                                        ...   \n",
       "21508                                 0.023451   \n",
       "21509                                 0.023451   \n",
       "21510                                 0.023451   \n",
       "21511                                 0.023451   \n",
       "21512                                 0.023451   \n",
       "\n",
       "       tr_rosetta_graph_degree_centrality  tr_rosetta_graph_load_centrality  \\\n",
       "0                                0.017544                          0.111866   \n",
       "1                                0.017544                          0.111866   \n",
       "2                                0.017544                          0.111866   \n",
       "3                                0.017544                          0.111866   \n",
       "4                                0.017544                          0.111866   \n",
       "...                                   ...                               ...   \n",
       "21508                            0.041825                          0.025188   \n",
       "21509                            0.041825                          0.025188   \n",
       "21510                            0.041825                          0.025188   \n",
       "21511                            0.041825                          0.025188   \n",
       "21512                            0.041825                          0.025188   \n",
       "\n",
       "       tr_rosetta_graph_harmonic_centrality tr_rosetta_graph_clustering  \\\n",
       "0                                   0.42778                    0.500000   \n",
       "1                                   0.42778                    0.500000   \n",
       "2                                   0.42778                    0.500000   \n",
       "3                                   0.42778                    0.500000   \n",
       "4                                   0.42778                    0.500000   \n",
       "...                                     ...                         ...   \n",
       "21508                               0.69352                    0.490909   \n",
       "21509                               0.69352                    0.490909   \n",
       "21510                               0.69352                    0.490909   \n",
       "21511                               0.69352                    0.490909   \n",
       "21512                               0.69352                    0.490909   \n",
       "\n",
       "      dssp_sec_struct  dssp_rsa dssp_asa  dssp_phi  dssp_psi  \n",
       "0                 NaN       NaN      NaN       NaN       NaN  \n",
       "1                 NaN       NaN      NaN       NaN       NaN  \n",
       "2                 NaN       NaN      NaN       NaN       NaN  \n",
       "3                 NaN       NaN      NaN       NaN       NaN  \n",
       "4                 NaN       NaN      NaN       NaN       NaN  \n",
       "...               ...       ...      ...       ...       ...  \n",
       "21508               H  0.072595      8.0     -55.2     -52.1  \n",
       "21509               H  0.072595      8.0     -55.2     -52.1  \n",
       "21510               H  0.072595      8.0     -55.2     -52.1  \n",
       "21511               H  0.072595      8.0     -55.2     -52.1  \n",
       "21512               H  0.072595      8.0     -55.2     -52.1  \n",
       "\n",
       "[21513 rows x 78 columns]"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "features = [\n",
    "        \"aa1\",\n",
    "        \"aa2\",\n",
    "        \"ev_frequency\",\n",
    "        \"ev_conservation\",\n",
    "        \"ev_independent\",\n",
    "        \"ev_epistatic\",\n",
    "        \"netsurf_rsa\",\n",
    "        \"netsurf_asa\",\n",
    "        \"netsurf_p_q3_H\",\n",
    "        \"netsurf_p_q3_E\",\n",
    "        \"netsurf_p_q3_C\",\n",
    "        \"netsurf_p_q8_G\",\n",
    "        \"netsurf_p_q8_H\",\n",
    "        \"netsurf_p_q8_I\",\n",
    "        \"netsurf_p_q8_B\",\n",
    "        \"netsurf_p_q8_E\",\n",
    "        \"netsurf_p_q8_S\",\n",
    "        \"netsurf_p_q8_T\",\n",
    "        \"netsurf_p_q8_C\",\n",
    "        \"netsurf_phi\",\n",
    "        \"netsurf_psi\",\n",
    "        \"netsurf_disorder\",\n",
    "        \"hmm_pssm_A\",\n",
    "        \"hmm_pssm_C\",\n",
    "        \"hmm_pssm_D\",\n",
    "        \"hmm_pssm_E\",\n",
    "        \"hmm_pssm_F\",\n",
    "        \"hmm_pssm_G\",\n",
    "        \"hmm_pssm_H\",\n",
    "        \"hmm_pssm_I\",\n",
    "        \"hmm_pssm_K\",\n",
    "        \"hmm_pssm_L\",\n",
    "        \"hmm_pssm_M\",\n",
    "        \"hmm_pssm_N\",\n",
    "        \"hmm_pssm_P\",\n",
    "        \"hmm_pssm_Q\",\n",
    "        \"hmm_pssm_R\",\n",
    "        \"hmm_pssm_S\",\n",
    "        \"hmm_pssm_T\",\n",
    "        \"hmm_pssm_V\",\n",
    "        \"hmm_pssm_W\",\n",
    "        \"hmm_pssm_Y\",\n",
    "        \"hmm_pssm_aa1_likelyhood\",\n",
    "        \"hmm_pssm_aa2_likelyhood\",\n",
    "        \"hmm_pssm_delta_likelyhood\",\n",
    "        \"tr_rosetta_graph_closeness_centrality\",\n",
    "        \"tr_rosetta_graph_betweenness_centrality\",\n",
    "        \"tr_rosetta_graph_degree_centrality\",\n",
    "        \"tr_rosetta_graph_load_centrality\",\n",
    "        \"tr_rosetta_graph_harmonic_centrality\",\n",
    "        \"tr_rosetta_graph_clustering\",\n",
    "    ]\n",
    "\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "thrown-liverpool",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>aa1</th>\n",
       "      <th>aa2</th>\n",
       "      <th>ev_frequency</th>\n",
       "      <th>ev_conservation</th>\n",
       "      <th>ev_independent</th>\n",
       "      <th>ev_epistatic</th>\n",
       "      <th>netsurf_rsa</th>\n",
       "      <th>netsurf_asa</th>\n",
       "      <th>netsurf_p_q3_H</th>\n",
       "      <th>netsurf_p_q3_E</th>\n",
       "      <th>...</th>\n",
       "      <th>hmm_pssm_Y</th>\n",
       "      <th>hmm_pssm_aa1_likelyhood</th>\n",
       "      <th>hmm_pssm_aa2_likelyhood</th>\n",
       "      <th>hmm_pssm_delta_likelyhood</th>\n",
       "      <th>tr_rosetta_graph_closeness_centrality</th>\n",
       "      <th>tr_rosetta_graph_betweenness_centrality</th>\n",
       "      <th>tr_rosetta_graph_degree_centrality</th>\n",
       "      <th>tr_rosetta_graph_load_centrality</th>\n",
       "      <th>tr_rosetta_graph_harmonic_centrality</th>\n",
       "      <th>tr_rosetta_graph_clustering</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dms_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>E1_Ubiquitin</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>57</td>\n",
       "      <td>57</td>\n",
       "      <td>57</td>\n",
       "      <td>57</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PSD95pdz3</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pab1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>49</td>\n",
       "      <td>49</td>\n",
       "      <td>49</td>\n",
       "      <td>49</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ubiquitin</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>110</td>\n",
       "      <td>110</td>\n",
       "      <td>110</td>\n",
       "      <td>110</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WW_domain</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>beta-lactamase</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>569</td>\n",
       "      <td>569</td>\n",
       "      <td>569</td>\n",
       "      <td>569</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gb1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hsp90</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>464</td>\n",
       "      <td>464</td>\n",
       "      <td>464</td>\n",
       "      <td>464</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>kka2_1:2</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>625</td>\n",
       "      <td>625</td>\n",
       "      <td>625</td>\n",
       "      <td>625</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9 rows Ã— 51 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                aa1  aa2  ev_frequency  ev_conservation  ev_independent  \\\n",
       "dms_id                                                                    \n",
       "E1_Ubiquitin      0    0            57               57              57   \n",
       "PSD95pdz3         0    0             0                0               0   \n",
       "Pab1              0    0            49               49              49   \n",
       "Ubiquitin         0    0           110              110             110   \n",
       "WW_domain         0    0            10               10              10   \n",
       "beta-lactamase    0    0           569              569             569   \n",
       "gb1               0    0             0                0               0   \n",
       "hsp90             0    0           464              464             464   \n",
       "kka2_1:2          0    0           625              625             625   \n",
       "\n",
       "                ev_epistatic  netsurf_rsa  netsurf_asa  netsurf_p_q3_H  \\\n",
       "dms_id                                                                   \n",
       "E1_Ubiquitin              57            0            0               0   \n",
       "PSD95pdz3                  0            0            0               0   \n",
       "Pab1                      49            0            0               0   \n",
       "Ubiquitin                110            0            0               0   \n",
       "WW_domain                 10            0            0               0   \n",
       "beta-lactamase           569            0            0               0   \n",
       "gb1                        0            0            0               0   \n",
       "hsp90                    464            0            0               0   \n",
       "kka2_1:2                 625            0            0               0   \n",
       "\n",
       "                netsurf_p_q3_E  ...  hmm_pssm_Y  hmm_pssm_aa1_likelyhood  \\\n",
       "dms_id                          ...                                        \n",
       "E1_Ubiquitin                 0  ...           0                        0   \n",
       "PSD95pdz3                    0  ...           0                        0   \n",
       "Pab1                         0  ...           0                        0   \n",
       "Ubiquitin                    0  ...           0                        0   \n",
       "WW_domain                    0  ...           0                        0   \n",
       "beta-lactamase               0  ...           0                        0   \n",
       "gb1                          0  ...           0                        0   \n",
       "hsp90                        0  ...           0                        0   \n",
       "kka2_1:2                     0  ...           0                        0   \n",
       "\n",
       "                hmm_pssm_aa2_likelyhood  hmm_pssm_delta_likelyhood  \\\n",
       "dms_id                                                               \n",
       "E1_Ubiquitin                          0                          0   \n",
       "PSD95pdz3                             0                          0   \n",
       "Pab1                                  0                          0   \n",
       "Ubiquitin                             0                          0   \n",
       "WW_domain                             0                          0   \n",
       "beta-lactamase                        0                          0   \n",
       "gb1                                   0                          0   \n",
       "hsp90                                 0                          0   \n",
       "kka2_1:2                              0                          0   \n",
       "\n",
       "                tr_rosetta_graph_closeness_centrality  \\\n",
       "dms_id                                                  \n",
       "E1_Ubiquitin                                        0   \n",
       "PSD95pdz3                                           0   \n",
       "Pab1                                                0   \n",
       "Ubiquitin                                           0   \n",
       "WW_domain                                           0   \n",
       "beta-lactamase                                      0   \n",
       "gb1                                                 0   \n",
       "hsp90                                               0   \n",
       "kka2_1:2                                            0   \n",
       "\n",
       "                tr_rosetta_graph_betweenness_centrality  \\\n",
       "dms_id                                                    \n",
       "E1_Ubiquitin                                          0   \n",
       "PSD95pdz3                                             0   \n",
       "Pab1                                                  0   \n",
       "Ubiquitin                                             0   \n",
       "WW_domain                                             0   \n",
       "beta-lactamase                                        0   \n",
       "gb1                                                   0   \n",
       "hsp90                                                 0   \n",
       "kka2_1:2                                              0   \n",
       "\n",
       "                tr_rosetta_graph_degree_centrality  \\\n",
       "dms_id                                               \n",
       "E1_Ubiquitin                                     0   \n",
       "PSD95pdz3                                        0   \n",
       "Pab1                                             0   \n",
       "Ubiquitin                                        0   \n",
       "WW_domain                                        0   \n",
       "beta-lactamase                                   0   \n",
       "gb1                                              0   \n",
       "hsp90                                            0   \n",
       "kka2_1:2                                         0   \n",
       "\n",
       "                tr_rosetta_graph_load_centrality  \\\n",
       "dms_id                                             \n",
       "E1_Ubiquitin                                   0   \n",
       "PSD95pdz3                                      0   \n",
       "Pab1                                           0   \n",
       "Ubiquitin                                      0   \n",
       "WW_domain                                      0   \n",
       "beta-lactamase                                 0   \n",
       "gb1                                            0   \n",
       "hsp90                                          0   \n",
       "kka2_1:2                                       0   \n",
       "\n",
       "                tr_rosetta_graph_harmonic_centrality  \\\n",
       "dms_id                                                 \n",
       "E1_Ubiquitin                                       0   \n",
       "PSD95pdz3                                          0   \n",
       "Pab1                                               0   \n",
       "Ubiquitin                                          0   \n",
       "WW_domain                                          0   \n",
       "beta-lactamase                                     0   \n",
       "gb1                                                0   \n",
       "hsp90                                              0   \n",
       "kka2_1:2                                           0   \n",
       "\n",
       "                tr_rosetta_graph_clustering  \n",
       "dms_id                                       \n",
       "E1_Ubiquitin                              0  \n",
       "PSD95pdz3                                 0  \n",
       "Pab1                                      0  \n",
       "Ubiquitin                                 0  \n",
       "WW_domain                                 0  \n",
       "beta-lactamase                            0  \n",
       "gb1                                       0  \n",
       "hsp90                                     0  \n",
       "kka2_1:2                                  0  \n",
       "\n",
       "[9 rows x 51 columns]"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "na_count = df[features].isna()\n",
    "dms_id = df[['dms_id']]\n",
    "dms_id.join(na_count).groupby('dms_id').sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "typical-initial",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "aa1                                           0\n",
       "aa2                                           0\n",
       "ev_frequency                               1884\n",
       "ev_conservation                            1884\n",
       "ev_independent                             1884\n",
       "ev_epistatic                               1884\n",
       "netsurf_rsa                                   0\n",
       "netsurf_asa                                   0\n",
       "netsurf_p_q3_H                                0\n",
       "netsurf_p_q3_E                                0\n",
       "netsurf_p_q3_C                                0\n",
       "netsurf_p_q8_G                                0\n",
       "netsurf_p_q8_H                                0\n",
       "netsurf_p_q8_I                                0\n",
       "netsurf_p_q8_B                                0\n",
       "netsurf_p_q8_E                                0\n",
       "netsurf_p_q8_S                                0\n",
       "netsurf_p_q8_T                                0\n",
       "netsurf_p_q8_C                                0\n",
       "netsurf_phi                                   0\n",
       "netsurf_psi                                   0\n",
       "netsurf_disorder                              0\n",
       "hmm_pssm_A                                    0\n",
       "hmm_pssm_C                                    0\n",
       "hmm_pssm_D                                    0\n",
       "hmm_pssm_E                                    0\n",
       "hmm_pssm_F                                    0\n",
       "hmm_pssm_G                                    0\n",
       "hmm_pssm_H                                    0\n",
       "hmm_pssm_I                                    0\n",
       "hmm_pssm_K                                    0\n",
       "hmm_pssm_L                                    0\n",
       "hmm_pssm_M                                    0\n",
       "hmm_pssm_N                                    0\n",
       "hmm_pssm_P                                    0\n",
       "hmm_pssm_Q                                    0\n",
       "hmm_pssm_R                                    0\n",
       "hmm_pssm_S                                    0\n",
       "hmm_pssm_T                                    0\n",
       "hmm_pssm_V                                    0\n",
       "hmm_pssm_W                                    0\n",
       "hmm_pssm_Y                                    0\n",
       "hmm_pssm_aa1_likelyhood                       0\n",
       "hmm_pssm_aa2_likelyhood                       0\n",
       "hmm_pssm_delta_likelyhood                     0\n",
       "tr_rosetta_graph_closeness_centrality         0\n",
       "tr_rosetta_graph_betweenness_centrality       0\n",
       "tr_rosetta_graph_degree_centrality            0\n",
       "tr_rosetta_graph_load_centrality              0\n",
       "tr_rosetta_graph_harmonic_centrality          0\n",
       "tr_rosetta_graph_clustering                   0\n",
       "dtype: int64"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "na_count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "novel-russian",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
